js_modules/dagit/packages/core/src/runs/ReexecutionDialog.tsx

```diff
@@ -0,0 +1,343 @@
+  if (!error) {
+    return 'Unknown error';
+  }
+  switch (error.__typename) {
+    case 'ConflictingExecutionParamsError':
+      return 'Conflicting execution parameters';
+    case 'InvalidOutputError':
+      return 'Invalid output';
+    case 'InvalidStepError':
+      return 'Invalid step';
+    case 'NoModeProvidedError':
+      return 'No mode provided';
+    case 'PipelineNotFoundError':
+      return 'Job not found in workspace';
+    case 'PresetNotFoundError':
+      return 'Preset not found';
+    case 'PythonError':
+      return error.message;
+    case 'RunConfigValidationInvalid':
+      return 'Run config invalid';
+    case 'RunConflict':
+      return 'Run conflict';
+    case 'UnauthorizedError':
+      return 'Re-execution not authorized';
+    default:
+      return 'Unknown error';
+  }
+};
+
+export type ReexecutionState = {completed: number; errors: {[id: string]: Error}};
+
+type ReexecutionDialogState = {
+  frozenRuns: SelectedRuns;
+  step: 'initial' | 'reexecuting' | 'completed';
+  reexecution: ReexecutionState;
+};
+
+type SelectedRuns = {[id: string]: RunTableRunFragment};
+
+const initializeState = (selectedRuns: SelectedRuns): ReexecutionDialogState => {
+  return {
+    frozenRuns: selectedRuns,
+    step: 'initial',
+    reexecution: {completed: 0, errors: {}},
+  };
+};
+
+type ReexecutionDialogAction =
+  | {type: 'reset'; frozenRuns: SelectedRuns}
+  | {type: 'start'}
+  | {type: 'reexecution-success'}
+  | {type: 'reexecution-error'; id: string; error: Error}
+  | {type: 'complete'};
+
+const reexecutionDialogReducer = (
+  prevState: ReexecutionDialogState,
+  action: ReexecutionDialogAction,
+): ReexecutionDialogState => {
+  switch (action.type) {
+    case 'reset':
+      return initializeState(action.frozenRuns);
+    case 'start':
+      return {...prevState, step: 'reexecuting'};
+    case 'reexecution-success': {
+      const {reexecution} = prevState;
+      return {
+        ...prevState,
+        step: 'reexecuting',
+        reexecution: {...reexecution, completed: reexecution.completed + 1},
+      };
+    }
+    case 'reexecution-error': {
+      const {reexecution} = prevState;
+      return {
+        ...prevState,
+        step: 'reexecuting',
+        reexecution: {
+          ...reexecution,
+          completed: reexecution.completed + 1,
+          errors: {...reexecution.errors, [action.id]: action.error},
+        },
+      };
+    }
+    case 'complete':
+      return {...prevState, step: 'completed'};
+  }
+};
+
+export const ReexecutionDialog = (props: Props) => {
+  const {isOpen, onClose, onComplete, selectedRuns} = props;
+
+  // Freeze the selected IDs, since the list may change as runs continue processing and
+  // re-executing. We want to preserve the list we're given.
+  const frozenRuns = React.useRef<SelectedRuns>(selectedRuns);
+
+  const [state, dispatch] = React.useReducer(
+    reexecutionDialogReducer,
+    frozenRuns.current,
```

Comment:
It doesn't seem like any of the state transitions modify this value so just using `frozenRuns.current` might be better? It's a little tricky that some of the code on 137-147 uses the state version and some uses the ref version.

Reply:
It updates when the dialog transitions from closed to open. We could update the ref when this happens, but we'd also want to re-render the dialog with the new list of runs, so IIRC that's why it's using a `dispatch`.